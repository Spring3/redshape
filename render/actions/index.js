import issueActions from './issue.actions';
import timeEntryActions from './timeEntry.actions';
import settingsActions from './settings.actions';

export default {
  issue: issueActions,
  timeEntry: timeEntryActions,
  settings: settingsActions
};
