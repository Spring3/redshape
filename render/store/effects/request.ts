import { ipcRenderer, remote } from 'electron';
import { Response } from '../../../types';

const crypto = remote.require('crypto');

type Config = {
  endpoint: string;
  token?: string;
}

type QueryConfig = {
  payload: Record<string, any>;
  config?: Config;
}

const query = (config: QueryConfig): Promise<Response> => {
  const id = crypto.randomBytes(10).toString('hex');

  return new Promise((resolve) => {
    ipcRenderer.send('request', JSON.stringify({ ...config, id }));

    ipcRenderer.once(`response:${id}`, (event, response) => {
      console.log('response:', response);
      resolve(response);
    });
  });
};

export {
  query,
};

export type {
  QueryConfig
};
