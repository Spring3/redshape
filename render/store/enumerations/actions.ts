import type { IAction, Context } from 'overmind';
import { Activity, Response } from '../../../types';

const fetchActivities: IAction<void, Promise<Response<Activity[]>>> = async ({ effects, state }: Context) => {
  const response = await effects.mainProcess.request({
    payload: {
      method: 'GET',
      route: 'enumerations/time_entry_activities.json',
    }
  });

  if (response.success) {
    state.enumerations.activities = response.data.activities as Activity[] ?? [];
    return {
      success: true,
      data: state.enumerations.activities,
    };
  }

  return {
    success: false,
    error: response.error
  };
};

export {
  fetchActivities
};
