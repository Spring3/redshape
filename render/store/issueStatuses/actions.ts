import { IAction, Context } from 'overmind';
import { IssueStatus, PaginatedActionResponse } from '../../../types';
import { indexById } from '../helpers';

const getAll: IAction<void, Promise<PaginatedActionResponse<IssueStatus>>> = async ({ effects, state }: Context) => {
  const response = await effects.mainProcess.request({
    payload: {
      method: 'GET',
      route: 'issue_statuses.json'
    }
  });

  if (response.success) {
    state.issueStatuses.byId = {
      ...state.issueStatuses,
      ...indexById(response.payload.issueStatuses)
    };

    return {
      success: true,
      data: {
        items: response.payload.issueStatuses,
        total: response.payload.total,
        limit: response.payload.limit,
        offset: response.payload.offset
      },
      hasMore: response.payload.total > (response.payload.offset + response.payload.issueStatuses.length)
    };
  }

  return {
    success: false,
    data: {
      items: [],
      total: 0,
      limit: response.payload.limit,
      offset: response.payload.offset
    },
    hasMore: false,
    error: response.error
  };
};

export {
  getAll
};
