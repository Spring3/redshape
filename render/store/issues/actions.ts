import type { Context, IAction } from 'overmind';
import { Issue, PaginatedActionResponse, Response } from '../../../types';
import { indexById } from '../helpers';

type GetManyIssuesArgs = {
  filters: {
    assigned_to_id?: number | string;
    status_id?: '*' | 'open' | 'closed';
    subject?: string;
    sort?: string;
  },
  offset?: number;
  limit?: number;
}

const getMany: IAction<GetManyIssuesArgs, Promise<PaginatedActionResponse<Issue>>> = async ({ effects, state }: Context, { filters, offset = 0, limit = 20 }) => {
  const query = {
    limit: limit ? Math.abs(limit) : undefined,
    offset: offset ? Math.abs(offset) : 0,
    include: 'attachments,relations',
    ...filters
  };

  const response = await effects.mainProcess.request({
    payload: {
      route: 'issues.json',
      method: 'GET',
      query
    }
  });

  if (response.success) {
    state.issues.byId = {
      ...state.issues.byId,
      ...indexById(response.payload.issues)
    };

    console.log('payload', response.payload);

    return {
      success: true,
      data: {
        items: response.payload.issues,
        total: response.payload.total,
        limit: response.payload.limit,
        offset: response.payload.offset
      },
      hasMore: response.payload.total > (response.payload.offset + response.payload.issues.length),
    };
  }

  return {
    success: false,
    data: {
      items: [],
      total: 0,
      limit,
      offset
    },
    hasMore: false,
    error: response.error
  };
};

type GetOneIssueArgs = {
  id: string;
}

const getOne: IAction<GetOneIssueArgs, Promise<Response<Issue>>> = async ({ effects, state }: Context, { id }) => {
  const response = await effects.mainProcess.request({
    payload: {
      route: `issues/${id}.json`,
      method: 'GET',
      query: {
        include: 'attachments,children,relations,journals',
      }
    }
  });

  if (response.success) {
    const issue = response.payload as Issue;
    state.issues.byId[id] = issue;
    return {
      success: true,
      data: issue
    };
  }

  return {
    success: false,
    error: response.error
  };
};

export {
  getOne,
  getMany
};
