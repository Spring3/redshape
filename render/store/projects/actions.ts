import type { IAction, Context } from 'overmind';
import { PaginatedActionResponse } from '../../../types';
import { indexById } from '../helpers';

type GetManyProjectsArgs = {
  offset?: number;
  limit?: number;
}

const getManyProjects: IAction<GetManyProjectsArgs, Promise<PaginatedActionResponse<any>>> = async ({ state, effects }: Context, { offset = 0, limit = 20 }) => {
  const normalizedOffset = offset ? Math.abs(offset) : 0;
  const normalizedLimit = limit ? Math.abs(limit) : undefined;

  const response = await effects.mainProcess.request({
    payload: {
      route: 'projects.json',
      method: 'GET',
      query: {
        offset: normalizedOffset,
        include: 'time_entry_activities',
        limit: normalizedLimit
      }
    }
  });

  if (response.success) {
    state.projects.byId = {
      ...state.projects.byId,
      ...indexById(response.data.projects)
    };

    console.log('payload', response.data);

    return {
      success: true,
      data: {
        items: response.data.projects,
        total: response.data.total,
        limit: response.data.limit,
        offset: response.data.offset
      },
      hasMore: response.data.total > response.data.projects.length
    };
  }

  return {
    success: false,
    data: {
      items: [],
      total: 0,
      limit: response.data.limit,
      offset: response.data.offset
    },
    hasMore: false,
    error: response.error
  };
};

export {
  getManyProjects
};
