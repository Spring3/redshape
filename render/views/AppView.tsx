import React, { useState, useCallback } from 'react';
import styled from 'styled-components';
import { useNavigate, Route, Routes } from 'react-router-dom';
import _get from 'lodash/get';
import moment from 'moment';

import { Navbar } from '../components/Navbar';
import { Timer } from '../components/Timer';
import { TimeEntryModal } from '../components/TimeEntryModal';
import DragArea from '../components/DragArea';

import { hoursToDuration } from '../datetime';
import { useOvermindActions, useOvermindState } from '../store';
import { NavbarContextProvider } from '../contexts/NavbarContext';
import { useFetchAll } from '../hooks/useFetchAll';
import { SummaryPage } from './AppViewPages/SummaryPage';
import IssueDetailsPage from './AppViewPages/IssueDetailsPage';
import { Project, Issue } from '../../types';

const Grid = styled.div`
  height: 100%;
  display: grid;
  grid-template-rows: 60px auto 80px;
  grid-template-columns: repeat(auto-fit, minmax(70px, 1fr));
`;

const Content = styled.div`
  grid-column: span 12;
  grid-row: 2 / -1;
`;

type AppViewProps = {
  // resetTimer: () => void;
}

const AppView = () => {
  const [activities, setActivities] = useState([]);
  const [showTimeEntryModal, setShowTimeEntryModal] = useState(false);
  const [timeEntry, setTimeEntry] = useState<any>(null);

  const navigate = useNavigate();
  const state = useOvermindState();
  const actions = useOvermindActions();

  const requestProjects = useCallback(({ limit, offset }) => actions.projects.getManyProjects({ limit, offset }), [actions.projects.getManyProjects]);
  const { items: projects, isFetching, error } = useFetchAll<Project>({ request: requestProjects });

  const onTrackingStop = ({ issue, recordedTime }: { issue: Issue, recordedTime: number }) => {
    const existingActivities = _get(projects[issue.project.id], 'activities', []);
    const hours = parseFloat((recordedTime / 3600000).toFixed(3));
    setActivities(existingActivities.map(({ id, name }: { id: string, name: string }) => ({ value: id, label: name })));
    setShowTimeEntryModal(true);
    setTimeEntry({
      activity: {},
      issue: {
        id: issue.id,
        name: issue.subject
      },
      hours,
      duration: hoursToDuration(hours),
      comments: '',
      project: {
        id: issue.project.id,
        name: issue.project.name
      },
      spent_on: moment().format('YYYY-MM-DD'),
      user: {
        id: state.users.currentUser?.id,
        name: `${state.users.currentUser?.firstName} ${state.users.currentUser?.lastName}`
      }
    });
  };

  const closeTimeEntryModal = () => {
    setShowTimeEntryModal(false);
    setTimeEntry(null);
    // resetTimer();
  };

  if (!state.users.currentUser) {
    navigate('../', { replace: true });
    return null;
  }

  return (
    <Grid>
      <DragArea />
      <NavbarContextProvider>
        <Navbar />
        <Content>
          <Routes>
            <Route path="/" element={<SummaryPage />} />
            <Route path="/:id" element={<IssueDetailsPage />} />
          </Routes>
          <Timer
            autoStart
            onStop={onTrackingStop}
          />
          <TimeEntryModal
            isOpen={showTimeEntryModal}
            activities={activities}
            isUserAuthor={true}
            timeEntry={timeEntry}
            initialVolatileContent={true}
            onClose={closeTimeEntryModal}
          />
        </Content>
      </NavbarContextProvider>
    </Grid>
  );
};

export { AppView };
