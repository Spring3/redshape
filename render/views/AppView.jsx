import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { connect } from 'react-redux';
import { Route, Redirect } from 'react-router-dom';
import _get from 'lodash/get';
import moment from 'moment';

import actions from '../actions';
import Navbar from '../components/Navbar';
import Timer from '../components/Timer';
import SummaryPage from './AppViewPages/SummaryPage';
import IssueDetailsPage from './AppViewPages/IssueDetailsPage';
import TimeEntryModal from '../components/TimeEntryModal';
import DragArea from '../components/DragArea';
import storage from '../../common/storage';

const Grid = styled.div`
  height: 100%;
  display: grid;
  grid-template-rows: 60px auto 80px;
  grid-template-columns: repeat(auto-fit, minmax(70px, 1fr));
`;

const Content = styled.div`
  grid-column: span 12;
  grid-row: 2 / -1;
`;

class AppView extends Component {
  constructor(props) {
    super(props);

    this.state = {
      activities: [],
      showTimeEntryModal: false,
      timeEntry: null
    };
  }

  componentWillMount() {
    this.props.getProjectData();
  }

  onTrackingStop = (value, trackedIssue) => {
    const { userId, userName, projects } = this.props;
    const activities = _get(projects[trackedIssue.project.id], 'activities', []);
    this.setState({
      activities: activities.map(({ id, name }) => ({ value: id, label: name })),
      showTimeEntryModal: true,
      timeEntry: {
        activity: {},
        issue: {
          id: trackedIssue.id,
          name: trackedIssue.subject
        },
        hours: (value / 3600000).toFixed(2),
        comments: '',
        project: {
          id: trackedIssue.project.id,
          name: trackedIssue.project.name
        },
        spent_on: moment().format('YYYY-MM-DD'),
        user: {
          id: userId,
          name: userName
        }
      }
    });
    storage.delete('time_tracking');
  }

  closeTimeEntryModal = () => {
    this.setState({ showTimeEntryModal: false, timeEntry: null });
    this.props.resetTimer();
  }

  render() {
    const { showTimeEntryModal, timeEntry, activities } = this.state;
    const { userId, api_key, match } = this.props;

    return (
      <Grid>
        <DragArea />
        { (!userId || !api_key) ? (<Redirect to="/" />) : null }
        <Navbar />
        <Content>
          <Route exact path={`${match.path}/summary`} component={props => <SummaryPage {...props}/>} />
          <Route path={`${match.path}/issue/:id`} component={props => <IssueDetailsPage {...props}/>} />
          <Timer
            onStop={this.onTrackingStop}
            history={this.props.history}
          />
          <TimeEntryModal
            isOpen={showTimeEntryModal}
            isEditable={true}
            onClose={this.closeTimeEntryModal}
            activities={activities}
            isUserAuthor={true}
            timeEntry={timeEntry}
            onClose={this.closeTimeEntryModal}
          />
        </Content>
      </Grid>
    );
  }
}

AppView.propTypes = {
  userId: PropTypes.number.isRequired,
  userName: PropTypes.string.isRequired,
  api_key: PropTypes.string.isRequired,
  match: PropTypes.shape({
    path: PropTypes.string.isRequired
  }).isRequired,
  logout: PropTypes.func.isRequired,
  resetTimer: PropTypes.func.isRequired,
  getProjectData: PropTypes.func.isRequired,
  projects: PropTypes.shape({
    activities: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired
    }).isRequired).isRequired
  }).isRequired
};

const mapStateToProps = state => ({
  userId: state.user.id,
  userName: state.user.name,
  api_key: state.user.api_key,
  projects: state.projects.data
});

const mapDispatchToProps = dispatch => ({
  logout: () => dispatch(actions.user.logout()),
  getProjectData: () => dispatch(actions.projects.getAll()),
  resetTimer: () => dispatch(actions.tracking.trackingReset())
});

export default connect(mapStateToProps, mapDispatchToProps)(AppView);
