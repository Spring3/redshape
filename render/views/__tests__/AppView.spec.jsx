import React from 'react';
import { Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import MockAdapter from 'axios-mock-adapter';
import {
  render, cleanup, fireEvent, act
} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import configureStore from 'redux-mock-store';
import { ThemeProvider } from 'styled-components';

import { PROJECT_GET_ALL } from '../../actions/project.actions';
import { TRACKING_RESET } from '../../actions/tracking.actions';
import * as axios from '../../../common/request';
import { theme } from '../../theme';
import AppView from '../AppView';

jest.mock('electron-store');

const mockStore = configureStore([thunk]);
const redmineEndpoint = 'redmine.test.test';
const token = 'multipass';

let axiosMock;
let getInstanceStub;

describe('AppView', () => {
  beforeAll(() => {
    axios.reset();
    axios.initialize(redmineEndpoint, token);
    const stub = axios.getInstance();
    getInstanceStub = jest.spyOn(axios, 'getInstance').mockReturnValue(stub);
    axiosMock = new MockAdapter(stub);
  });

  afterEach(() => {
    axiosMock.reset();
    cleanup();
  });

  afterAll(() => {
    axiosMock.restore();
    getInstanceStub.restore();
  });

  it('should fetch projects data on mount', () => {
    axiosMock.onGet('/projects.json').reply(() => Promise.resolve([200, {
      projects: [],
      totalCount: 10,
      offset: 0,
      limit: 10
    }]));
    const store = mockStore({
      user: {
        id: 1,
        name: 'John Wayne',
        redmineEndpoint: 'https://redmine.domain',
        api_key: '123abc'
      },
      issues: {
        all: {
          data: []
        }
      },
      projects: {
        data: {
          activities: [
            { id: 1, name: 'Activity 1' },
            { id: 2, name: 'Activity 2' }
          ]
        }
      },
      tracking: {
        isEnabled: false,
        issue: {
          subject: 'Test issue'
        }
      },
      timeEntry: {
        isFetching: false,
        error: undefined
      },
      settings: {
        idleBehavior: 0,
        discardIdleTime: true,
        advancedTimerControls: false,
        progressWithStep1: false
      }
    });

    const historyMock = {
      location: {
        pathname: '/app'
      },
      listen: () => () => { /* noop */ },
      push: jest.fn(),
      createHref: () => {
        historyMock.push('/');
        window.location.path = '/';
      }
    };

    render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <Router history={historyMock}>
            <AppView
              match={{
                isExact: false,
                params: {},
                path: undefined,
                url: '/app'
              }}
            />
          </Router>
        </ThemeProvider>
      </Provider>
    );

    expect(store.getActions().length).toBeGreaterThan(0);
    expect(store.getActions()[0].type).toBe(PROJECT_GET_ALL);
  });

  it('should logout when a user clicks the button', () => {
    axiosMock.onGet('/projects.json').reply(() => Promise.resolve([200, {
      projects: [],
      totalCount: 10,
      offset: 0,
      limit: 10
    }]));

    const store = mockStore({
      user: {
        id: 1,
        name: 'John Wayne',
        redmineEndpoint: 'https://redmine.domain',
        api_key: '123abc'
      },
      issues: {
        all: {
          data: []
        }
      },
      projects: {
        data: {
          activities: [
            { id: 1, name: 'Activity 1' },
            { id: 2, name: 'Activity 2' }
          ]
        }
      },
      tracking: {
        isEnabled: false,
        issue: {
          subject: 'Test issue'
        }
      },
      timeEntry: {
        isFetching: false,
        error: undefined
      },
      settings: {
        idleBehavior: 0,
        discardIdleTime: true,
        advancedTimerControls: false,
        progressWithStep1: false
      }
    });

    const historyMock = {
      location: {
        pathname: '/app'
      },
      listen: () => () => { /* noop */ },
      push: jest.fn(),
      createHref: () => {
        historyMock.push('/');
        window.location.path = '/';
      }
    };

    render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <Router history={historyMock}>
            <AppView
              match={{
                isExact: false,
                params: {},
                path: undefined,
                url: '/app'
              }}
            />
          </Router>
        </ThemeProvider>
      </Provider>
    );

    const signoutButton = document.querySelector('#signout');
    expect(signoutButton).toBeDefined();

    act(() => {
      fireEvent.click(signoutButton);
    });

    expect(historyMock.push).toHaveBeenCalledWith('/');
    const actions = store.getActions();
    expect(actions.length).toBeGreaterThan(0);
  });

  it('should open the modal with a new timeEntry if timer stops and wipe the storage', (done) => {
    axiosMock.onGet('/projects.json').reply(() => Promise.resolve([200, {
      projects: [],
      totalCount: 10,
      offset: 0,
      limit: 10
    }]));

    const state = {
      user: {
        id: 1,
        name: 'John Wayne',
        redmineEndpoint: 'https://redmine.domain',
        api_key: '123abc'
      },
      issues: {
        all: {
          data: []
        }
      },
      projects: {
        data: {
          1: {
            activities: [
              { id: 1, name: 'Activity 1' },
              { id: 2, name: 'Activity 2' }
            ]
          }
        }
      },
      tracking: {
        isEnabled: false,
        ms: 41000,
        issue: {
          id: 321,
          subject: 'Test issue',
          project: {
            id: 1,
            name: 'Project 1'
          }
        }
      },
      timeEntry: {
        isFetching: false,
        error: undefined
      },
      settings: {
        idleBehavior: 0,
        discardIdleTime: true,
        advancedTimerControls: false,
        progressWithStep1: false
      }
    };
    const store = mockStore(state);

    const historyMock = {
      location: {
        pathname: '/app'
      },
      listen: () => () => { /* noop */ },
      push: jest.fn(),
      createHref: () => {
        historyMock.push('/');
        window.location.path = '/';
      }
    };

    render(
      <Provider store={store}>
        <Router history={historyMock}>
          <ThemeProvider theme={theme}>
            <AppView
              match={{
                isExact: false,
                params: {},
                path: undefined,
                url: '/app'
              }}
            />
          </ThemeProvider>
        </Router>
      </Provider>
    );

    fireEvent.click(document.querySelector('#stop-timer'));
    fireEvent.click(document.querySelector('#btn-add'));
    setTimeout(() => {
      expect(
        store.getActions().find((action) => action.type === TRACKING_RESET)
      ).toBeDefined();
      done();
    }, 1);
  });
});
