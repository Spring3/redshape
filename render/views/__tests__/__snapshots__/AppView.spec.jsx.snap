// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppView should match the snapshot 1`] = `
.c4 {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #FF7079;
  -webkit-transition: color ease .2s;
  transition: color ease .2s;
}

.c4:hover {
  color: #EF6069;
}

.c4:hover svg {
  fill: #EF6069;
}

.c4:active,
.c4:focus,
.c4:visited {
  background: transparent;
}

.c1 {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  height: 23px;
  z-index: 3;
  -webkit-app-region: drag;
}

.c5 {
  color: #FF7079;
  -webkit-align-self: flex-end;
  -ms-flex-item-align: end;
  align-self: flex-end;
}

.c2 {
  position: fixed;
  top: 0px;
  left: 0px;
  right: 0px;
  z-index: 2;
  height: 50px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 20px 40px 0px 40px;
  background: linear-gradient(to bottom,#FFFFFF 85%,transparent);
}

.c2 ul {
  list-style-type: none;
  padding: 0;
  margin: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 ul li {
  display: inline;
  font-size: 15px;
  font-weight: bold;
  color: #3F3844;
  -webkit-transition: color ease .2s;
  transition: color ease .2s;
}

.c2 ul li a {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #3F3844;
  -webkit-transition: color ease .2s;
  transition: color ease .2s;
  padding-bottom: 5px;
}

.c2 ul li a.active {
  color: #FF7079;
  border-bottom: 2px solid #FF7079;
}

.c2 ul li:hover a {
  cursor: pointer;
  color: #FF7079;
}

.c2 ul:first-child li {
  margin: 0px 20px;
}

.c2 ul:first-child li:first-child {
  margin-left: 0;
}

.c2 ul:last-child li {
  margin: 0px 20px;
}

.c2 ul:last-child li:last-child {
  margin-right: 0px;
}

.c3 svg {
  border-radius: 3px;
  color: #FF7079;
  border: 2px solid transparent;
  -webkit-transition: all ease .2s;
  transition: all ease .2s;
}

.c3 svg:hover {
  color: #FF7079;
  border: 2px solid #FF7079;
}

.c3 svg {
  -webkit-animation: ikNrtx 2s ease-in infinite;
  animation: ikNrtx 2s ease-in infinite;
}

.c3 svg:hover {
  -webkit-animation-play-state: paused;
  animation-play-state: paused;
}

.c7 {
  -webkit-animation: gOHwoP .7s ease-in;
  animation: gOHwoP .7s ease-in;
  max-width: 100%;
  box-sizing: border-box;
  padding: 15px 20px;
  position: fixed;
  bottom: 0;
  width: 100%;
  background: #FFFFFF;
  display: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-shadow: 0px -2px 20px #F9F9F9;
  border-top: 2px solid #F9F9F9;
}

.c7 div.panel {
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  min-width: 520px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: 1800px;
}

.c7 div.buttons {
  margin: 0 20px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c7 div.time {
  margin: 0 20px;
  font-size: 16px;
  font-weight: bold;
}

.c7 div.buttons a {
  padding: 5px 0px;
}

.c7 div.buttons a:hover {
  background: #F9F9F9;
}

.c7 div.buttons a:first-child {
  margin-right: 20px;
}

.c7 div.buttons.buttons-advanced a {
  margin-right: 5px;
}

.c7 div.buttons.buttons-advanced a:last-child {
  margin-right: initial;
}

.c7 div.issueName {
  padding: 0 20px;
  max-width: 500px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c7 div.time {
  color: #FF7079;
}

.c7 input[name="comment"] {
  -webkit-box-flex: 2;
  -webkit-flex-grow: 2;
  -ms-flex-positive: 2;
  flex-grow: 2;
  margin-left: 20px;
  width: initial;
  border: none;
  border-radius: 0;
  border-bottom: 1px #F9F9F9 solid;
  color: #A4A4A4;
}

.c7 input[name="comment"]:focus {
  border: none;
  border-radius: 0;
  border-bottom: 1px #FF7079 solid;
  box-shadow: none;
}

.c8 {
  padding: 0px;
}

.c0 {
  height: 100%;
  display: grid;
  grid-template-rows: 60px auto 80px;
  grid-template-columns: repeat(auto-fit,minmax(70px,1fr));
}

.c6 {
  grid-column: span 12;
  grid-row: 2 / -1;
}

<div
  className="c0"
>
  <div
    className="c1"
  />
  <nav
    className="c2"
  >
    <ul>
      <li>
        <div>
          <a
            className="c3 c4"
            disabled={false}
            href="#"
            onClick={[Function]}
          >
            <svg
              className="mdi-icon "
              fill="currentColor"
              height={30}
              viewBox="0 0 24 24"
              width={30}
            >
              <path
                d="M20,11V13H8L13.5,18.5L12.08,19.92L4.16,12L12.08,4.08L13.5,5.5L8,11H20Z"
              />
            </svg>
          </a>
        </div>
      </li>
      <li>
        <a
          aria-current={null}
          href="#/app/summary/assigned"
          onClick={[Function]}
        >
          Assigned
        </a>
      </li>
      <li>
        <a
          aria-current={null}
          href="#/app/summary/author"
          onClick={[Function]}
        >
          Author
        </a>
      </li>
    </ul>
    <ul>
      <li>
        <a
          aria-current={null}
          href="#/app/settings"
          onClick={[Function]}
        >
          Settings
        </a>
      </li>
      <li>
        <span
          className="c5"
        >
          John Wayne
        </span>
      </li>
      <li>
        <a
          className="c4"
          disabled={false}
          href="#"
          id="signout"
          onClick={[Function]}
        >
          Sign out
        </a>
      </li>
    </ul>
  </nav>
  <div
    className="c6"
  >
    <div
      className="c7"
    >
      <div
        className="panel"
      >
        <div
          className="buttons"
        >
          <a
            className="c8 c4"
            disabled={false}
            href="#"
            onClick={[Function]}
          >
            <svg
              className="mdi-icon "
              fill="currentColor"
              height={35}
              viewBox="0 0 24 24"
              width={35}
            >
              <path
                d="M18,18H6V6H18V18Z"
              />
            </svg>
          </a>
          <a
            className="c8 c4"
            disabled={false}
            href="#"
            onClick={[Function]}
          >
            <svg
              className="mdi-icon "
              fill="currentColor"
              height={35}
              viewBox="0 0 24 24"
              width={35}
            >
              <path
                d="M14,19H18V5H14M6,19H10V5H6V19Z"
              />
            </svg>
          </a>
        </div>
        <div
          className="issueName"
        />
        <div
          className="time"
        >
          <span>
            00:00:00
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
